### ISSUE TITLE
YCbCr to RGB and vice versa color space conversion functions

### ISSUE THREAD

**Issue by 9prady9** [arrayfire/arrayfire#532]

**Comment by ghazanfar91** [comment]
May I suggest formulas for color space conversion?

**Comment by ghazanfar91** [comment]
RGB To YCbCr color conversion coefficients:

`float coeffs[12] = { 65.481f * scale, 128.553f * scale, 24.966f * scale, 16.0f * offset,
				-37.797f * scale, -74.203f * scale, 112.0f * scale, 128.0f * offset,
				112.0f * scale, -93.786f * scale, -18.214f * scale, 128.0f * offset};`

Value of `scale` and `offset` according to image type:

**8u** ---> `scale = 1.0f/255.0f`, `offset = 1.0f`
**16u** ---> `scale = 257.0f/65535.0f`, `offset = 257.0f`
**32f** ---> `scale = 1.0f/255.0f`, `offset = 1.0f/255.0f`

`float Y  = coeffs[0] * R + coeffs[1] * G + coeffs[2]  * B + coeffs[3];`
`float Cb = coeffs[4] * R + coeffs[5] * G + coeffs[6]  * B + coeffs[7];`
`float Cr = coeffs[8] * R + coeffs[9] * G + coeffs[10] * B + coeffs[11];`

**Comment by ghazanfar91** [comment]
YCbCr To RGB color conversion coefficients:

```
float coeffs[12] = 
{ 
    0.0045662f * scale,  0,  0.0062589f * scale, -0.8742024f * offset,
    0.0045662f * scale, -0.0015363f * scale, -0.0031881f * scale,  0.5316682f * offset,
    0.0045662f * scale,  0.0079107f * scale,  0 ,  -1.0856326f * offset
};
```

Value of scale and offset according to image type:

**8u** ---> `scale = 255.0f`, `offset = 255.0f`
**16u** ---> `scale = 65535.0f/257.0f`, `offset = 65535.0f`
**32f** ---> `scale = 255.0f`, `offset = 1.0f`

```
float R = coeffs[0] * Y + coeffs[1] * Cb + coeffs[2]  * Cr + coeffs[3];
float G = coeffs[4] * Y + coeffs[5] * Cb + coeffs[6]  * Cr + coeffs[7];
float B = coeffs[8] * Y + coeffs[9] * Cb + coeffs[10] * Cr + coeffs[11];
```

The above coefficients have been adapted from MATLAB file `ycbcr2rgb.m`. The values have been tested and compared against MATLAB results.

**Comment by ghazanfar91** [comment]
Assumed range of values:

**8u** -----> `0 - 255`
**16u** --->  `0 - 65535`
**32f** ---->  `0.0f - 1.0f`

**Comment by pavanky** [comment]
@ghazanfar91 Is there a academic paper you can cite for this ? We don't want to use values from proprietary implementations.

**Comment by ghazanfar91** [comment]
Well in the MATLAB's file a reference to a book is given from where these coefficients are taken:
```
C.A. Poynton, "A Technical Introduction to Digital Video", John Wiley & Sons, Inc., 1996, Chapter 9, Page 175`
```

I don't have a copy of it, but I can confirm the reference once I get hold of this book.

**Comment by pavanky** [comment]
@ghazanfar91 Thanks for pointing out the reference. We will see if we can find on our end as well.

**Comment by 9prady9** [comment]
The coefficients for RGB to YCbCr conversion are slightly off on wikipedia but they are accurate on the integer portion of the floating point values.

**Comment by 9prady9** [comment]
@pavanky @sgarizvi Would [this](http://www.fourcc.org/fccyvrgb.php) be a good source? the site provides crowd sourced information of media codecs and formats.

**Comment by pavanky** [comment]
They have citations, so I think it should be fine.

**Comment by 9prady9** [comment]
Fixes by #942
